KAFKA-18084: Added write locks in SharePartition where locks were async calls were being made

adixitconfluent opened the pull request:
### About
Added write locks in `SharePartition` where locks were async calls were being made and no locks were used. This could potentially lead to race condition when updating the state of offsets/batches in `SharePartition`.

### Testing
The code is tested with the help of already present unit and integration tests.

adixitconfluent committed:
Created branch

adixitconfluent committed:
Added locks where async calls are being made in share partition

adixitconfluent committed:
Trigger build

AndrewJSchofield reviewed (approved):
Thanks for the PR. I think this adds the appropriate locks in the `whenComplete` code. Taken with https://github.com/apache/kafka/pull/17965, I think https://issues.apache.org/jira/browse/KAFKA-18084 is addressed.

chia7712 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -389,79 +389,84 @@ public CompletableFuture<Void> maybeInitialize() {
+                    }
+                    // Set the partition state to Active and complete the future.
+                    partitionState = SharePartitionState.ACTIVE;
+                    future.complete(null);
```
It might be better not to complete the future while holding the write lock. The registered action for the `future` could potentially try to acquire another lock that is held by a different thread, which, in turn, could be waiting on the write lock - resulting in a deadlock.

In short, I prefer to narrow the scope of the write lock to avoid potential deadlock issues.

adixitconfluent replied:
Hi @chia7712 , should we not even complete futures when they are being completed with exceptions `completeInitializationWithException` within the write lock, for example - [here](https://github.com/apache/kafka/pull/17957/files#diff-83f8b0eae4289cefc39e38e4b85e64d49610ceb900eb456285f1ab8d240d9555R445) or IIUC, I should only move `future.complete(null)` within finally block after releasing the write lock and add an `if` statement that this should be performed only if future hasn't been completed yet.

chia7712 replied:
> should we not even complete futures when they are being completed with exceptions completeInitializationWithException within the write lock,

Yes, that should not hold the write lock too!

AndrewJSchofield replied:
I think that's going to cast the net a bit wider in this PR. I agree that it's a worthwhile thing to do.

apoorvmittal10 replied:
> The registered action for the future could potentially try to acquire another lock that is held by a different thread

Hmm, I don't see that should be a problem. There can anyways be multiple threads waiting on the same lock in other methods as well. Am I missing something here?

adixitconfluent replied:
@AndrewJSchofield @chia7712 @apoorvmittal10 I am reverting this change from this PR. Agree with @AndrewJSchofield , this change is a bit wider and doesn't align to the issues pointed in https://issues.apache.org/jira/browse/KAFKA-18084 . I'll make these change as part of issue https://issues.apache.org/jira/browse/KAFKA-18129 .

chia7712 replied:
> I don't see that should be a problem. There can anyways be multiple threads waiting on the same lock in other methods as well. Am I missing something here?

Yes, that’s a great question. One possible deadlock path from my perspective could occur as follows:

a_thread:

1. Holds the lock of a shared partition.
2. Attempts to complete a delayed share fetch.
3. Waits for the lock of the delayed share fetch.

b_thread:

1. Holds the lock of the delayed share fetch.
2. Attempts to access the shared partition.
3. Waits for the lock of the shared partition.

Based on these rules, the potential deadlock path could be as [this line](https://github.com/apache/kafka/blob/615f1a0bf999cd0b5ccf38c09e6afaba89538e9f/core/src/main/java/kafka/server/share/SharePartitionManager.java#L589). The PersisterStateManager thread acts as a_thread, while b_thread could be any thread attempting to complete a delayed share fetch while handling other types of requests.

At any rate, my point is that we should hold the lock only when absolutely necessary, even if we haven’t encountered an existing deadlock case yet. This minimizes the risk and improves overall robustness.

chia7712 replied:
> I'll make these change as part of issue https://issues.apache.org/jira/browse/KAFKA-18129 .

It seems KAFKA-18129 is intended to add a write lock to `maybeInitialize`. However, the additional discussion focuses on the appropriate scope of the write lock, particularly concerning the potential for deadlocks. If we can't reach a consensus on the lock scope, we could still add the write lock to `maybeInitialize` in this PR. It doesn't seem like a significant change to make at this point.

adixitconfluent replied:
Hi @chia7712 , I've reverted back the change to include write locks within persister.readState completion call in `maybeInitialize`

adixitconfluent replied:
Hi @chia7712 , maybe I am not understanding it correctly, so just want to clarify - In this example, https://github.com/apache/kafka/pull/17957#discussion_r1864856887 , a_thread holds the `lock.writeLock()` of a share partition, however b_thread will hold the fetch lock of the share partition. It won't hold the `lock.readLock()` of the share partition until the write lock by persister has been released. So how can there be a deadlock?
cc- @apoorvmittal10

chia7712 replied:
@adixitconfluent I haven't observed this scenario in the codebase, so the following discussion aims to prevent potential deadlocks in the future.

Let's assume we add an action that attempts to complete a delayed fetch to the future completed by `persister`. This means the `persister` thread holds the write lock of the shared partition and then requires the lock for the delayed fetch. Meanwhile, another thread holding the lock for the same delayed fetch needs to acquire the lock for the shared partition. This situation could lead to a deadlock based on my understanding.

adixitconfluent replied:
Thanks for explaining that. Yes, I agree this is a good future enhancement, **reiterating what you pointed out, https://issues.apache.org/jira/browse/KAFKA-18129 (this comment thread) does not cause any deadlock in the current code**. I've opened a PR for the same https://github.com/apache/kafka/pull/18053

chia7712 reviewed (commented):
@adixitconfluent thanks for this patch. I leave a major comment. PTAL

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -389,79 +389,84 @@ public CompletableFuture<Void> maybeInitialize() {
-                    completeInitializationWithException(future, exception);
-                    return;
-                }
+                lock.writeLock().lock();
```
Why not to release the lock after setting `partitionState = SharePartitionState.INITIALIZING;` and re-acquire later?

The method has double-checked lock where method `maybeCompleteInitialization(future)` is checked again to reduce lock overhead as the `EMPTY` state should be temporary at bootstrap. We should release the lock post setting the state and re-acquire when persister sends the response back. It shall be clean that way.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -389,79 +389,84 @@ public CompletableFuture<Void> maybeInitialize() {
-                    maybeUpdateCachedStateAndOffsets();
-                } else {
-                    endOffset = startOffset;
+                    stateEpoch = partitionData.stateEpoch();
+
+                    List<PersisterStateBatch> stateBatches = partitionData.stateBatches();
+                    for (PersisterStateBatch stateBatch : stateBatches) {
+                        if (stateBatch.firstOffset() < startOffset) {
+                            log.error("Invalid state batch found for the share partition: {}-{}. The base offset: {}"
+                                    + " is less than the start offset: {}.", groupId, topicIdPartition,
+                                stateBatch.firstOffset(), startOffset);
+                            completeInitializationWithException(future, new IllegalStateException(String.format("Failed to initialize the share partition %s-%s", groupId, topicIdPartition)));
+                            return;
+                        }
+                        InFlightBatch inFlightBatch = new InFlightBatch(EMPTY_MEMBER_ID, stateBatch.firstOffset(),
+                            stateBatch.lastOffset(), RecordState.forId(stateBatch.deliveryState()), stateBatch.deliveryCount(), null);
+                        cachedState.put(stateBatch.firstOffset(), inFlightBatch);
+                    }
+                    // Update the endOffset of the partition.
+                    if (!cachedState.isEmpty()) {
+                        // If the cachedState is not empty, findNextFetchOffset flag is set to true so that any AVAILABLE records
+                        // in the cached state are not missed
+                        findNextFetchOffset.set(true);
+                        endOffset = cachedState.lastEntry().getValue().lastOffset();
+                        // In case the persister read state RPC result contains no AVAILABLE records, we can update cached state
+                        // and start/end offsets.
+                        maybeUpdateCachedStateAndOffsets();
+                    } else {
+                        endOffset = startOffset;
+                    }
+                    // Set the partition state to Active and complete the future.
+                    partitionState = SharePartitionState.ACTIVE;
```
@adixitconfluent Just to be sure, there is no code change other that indentation here, correct?

adixitconfluent replied:
@apoorvmittal10 , yes, just the identation change. No code line modified here.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -1647,24 +1652,29 @@ void rollbackOrProcessStateUpdates(
+                    lock.writeLock().unlock();
+                }
             });
         } finally {
             lock.writeLock().unlock();
```
Again I would say to decouple the 2 locks and release the first one just after `updatedStates.forEach(state -> state.completeStateTransition(false));` as that only needs to be in lock.

apoorvmittal10 reviewed (commented):
Thanks for the PR, left couple of comments.

adixitconfluent committed:
Removed change for maybrInitialize

adixitconfluent committed:
Reverted the write lock change for maybeInitialize and addressed Apoorv's comment

adixitconfluent committed:
Addressed Apoorv's comment about maybeInitialize

chia7712 commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -490,11 +490,14 @@ public void testMaybeInitializeSharePartitionAgainConcurrentRequests() throws In
         for (CompletableFuture<Void> result : results) {
             assertTrue(result.isDone());
-            assertFalse(result.isCompletedExceptionally());
+            if (!result.isCompletedExceptionally())
```
Pardon me, this PR's primary purpose is to add the write lock to maybeInitialize. Why are these additional changes needed?

chia7712 replied:
The test appears to assume that all threads will successfully initialize the share partition concurrently. However, due to the narrowed lock scope introduced in this PR, some threads are now failing because the partition is in the `INITIALIZING` state.

chia7712 replied:
Maybe we can simplify the code by lambda?
```
        assertTrue(results.stream().allMatch(CompletableFuture::isDone));
        assertFalse(results.stream().allMatch(CompletableFuture::isCompletedExceptionally));
```

adixitconfluent replied:
Hi @chia7712 , yes, if not for the change, this test would become highly flaky.

> Maybe we can simplify the code by lambda?
> 
> ```
>         assertTrue(results.stream().allMatch(CompletableFuture::isDone));
>         assertFalse(results.stream().allMatch(CompletableFuture::isCompletedExceptionally));
> ```
Made this change. Please review the PR. Thanks!

adixitconfluent committed:
Trigger build

adixitconfluent committed:
Addressed Chia's comment around using lambda function in the changed test

chia7712 commented:
>  thanks for the approval, can you please merge the PR?

@apoorvmittal10 Do you have any feedback for this PR? Aside from the lock scope, all other changes are ready to be merged.

apoorvmittal10 commented:
> > thanks for the approval, can you please merge the PR?
> 
> @apoorvmittal10 Do you have any feedback for this PR? Aside from the lock scope, all other changes are ready to be merged.

Mine is more on the way the locks are acquired and release. There is no point to keep holding the locks when a async API has been called. @adixitconfluent are we saying to have that change in the next PR? If yes, then it seems fine to me.

adixitconfluent commented:
> > > thanks for the approval, can you please merge the PR?
> > 
> > 
> > @apoorvmittal10 Do you have any feedback for this PR? Aside from the lock scope, all other changes are ready to be merged.
> 
> Mine is more on the way the locks are acquired and release. There is no point to keep holding the locks when a async API has been called. @adixitconfluent are we saying to have that change in the next PR? If yes, then it seems fine to me.

Hi @apoorvmittal10 , i have addressed your comments where you mentioned to decouple the locks in both maybeInitialize and post persister write state RPC calls. @chia7712 is mentioning about this comment thread https://github.com/apache/kafka/pull/17957#discussion_r1862570482 where we've not reached consensus that we want to separate out the future completion from the lock scope. Other than that, all the changes are done. I would like to merge this PR since other comments have been resolved and it would be good if we can reach consensus on the former thread, so if needed, I can make the changes before the feature freeze.

apoorvmittal10 commented on core/src/main/java/kafka/server/share/SharePartition.java:
```diff
@@ -380,6 +380,14 @@ public CompletableFuture<Void> maybeInitialize() {
+        } finally {
+            lock.writeLock().unlock();
+        }
+        try {
```
Shouldn't this `try` be inside of the response from persister read state, why to have a lock while invoking the persister read call?

apoorvmittal10 replied:
Moreover is this try block needed? I don't think we should have it, there is no exception occurring.

adixitconfluent replied:
We need this `try` in case we have `persister.readState` throw  any exceptions. We even have a test case `testMaybeInitializeWithReadException` in `SharePartitionTest` that verifies this behaviour and it will start failing in case we get rid of this `try catch` block.

chia7712 replied:
> Shouldn't this try be inside of the response from persister read state, why to have a lock while invoking the persister read call?

The `try` block is not related to the lock; instead, it handles exceptions thrown by `persister.readState`. This introduces another inconsistency in error handling - The `persister.writeState` can also throw exceptions, but we do not use `try-catch` blocks to complete the `future` with an exception.

chia7712 replied:
We should follow up by unifying the error handling for all `persister` methods. Since `persister` uses `CompletableFuture` as returned type, we can leverage it to propagate exceptions instead of handling them separately in both `try-catch` and `whenComplete` blocks.

adixitconfluent replied:
@chia7712 , actually exceptions are getting handled when `persister.writeState` throws exceptions. We have a test case `testWriteShareGroupStateWithWriteException` in `SharePartitionTest` that verifies this beahviour.

adixitconfluent replied:
@chia7712 @apoorvmittal10 I have added a commit where I have added `try catch` to `persister.writeState` call in `SharePartition`.Now, the error handling is consistent in both `persister.writeState` and `persister.readState` calls. `try catch` block protects from `Runtime exceptions` and `whenComplete` exception handling handles all futures which are completed exceptionally. Also added a unit test as well.

chia7712 replied:
> actually exceptions are getting handled when persister.writeState throws exceptions. We have a test case testWriteShareGroupStateWithWriteException in SharePartitionTest that verifies this beahviour.

`testWriteShareGroupStateWithWriteException` verify the exception wrapped by `CompletableFuture`. However, `persister.writeState` could throw exception directly. The exception is not wrapped by `CompletableFuture` 

https://github.com/apache/kafka/blob/d00f0ecf1a1a082c97564f4b807e7a342472b57a/share/src/main/java/org/apache/kafka/server/share/persister/DefaultStatePersister.java#L82

apoorvmittal10 replied:
>We need this try in case we have persister.readState throw any exceptions

Hmmm, I was more in line with to have the try block when lock is required and not have `persister.readState/persister.writeState` in the try block as persister should return the completable future with exception, if failed, and should never throw an exception directly. The javadoc of persister do not specify throwing any exceptions directly.

>We even have a test case testMaybeInitializeWithReadException in SharePartitionTest that verifies this behaviour

We can change the expectation of the test in the 2nd case of the test. Also we should see the exception is being handled at top level if this happens but we can't write entire code in try/catch.

adixitconfluent replied:
@chia7712, agreed, my latest commit as mentioned on comment https://github.com/apache/kafka/pull/17957#discussion_r1866535510 contains the change where exceptions thrown directly are also handled now.

chia7712 replied:
> as persister should return the completable future with exception, if failed, and should never throw an exception directly. 

yes I love this style too :)

chia7712 replied:
@adixitconfluent I see your last update use the `try-catch` to handle exception thrown by `persister.readState/persister.writeState`. As @apoorvmittal10 comment and my previous comment (https://github.com/apache/kafka/pull/17957#discussion_r1866507046), maybe `persister` can leverage `CompletableFuture` to propagate the exception? WDYT?

adixitconfluent replied:
@chia7712, I can remove the `try catch` from `persister.readState/persister.writeState` calls and modify the test cases for the corresponding behaviour. From what I see in the persister code lines [here](https://github.com/apache/kafka/blob/trunk/share/src/main/java/org/apache/kafka/server/share/persister/DefaultStatePersister.java#L208) and [here](https://github.com/apache/kafka/blob/trunk/share/src/main/java/org/apache/kafka/server/share/persister/DefaultStatePersister.java#L118), `CompletableFuture` is being utilized to propagate the exception. I think the `try catch` in `persister.readState` was added as an additional check in the initial iterations of development, but we don't seem to need it now. If we're all aligned, then  I'll make this change and re-raise a review.

chia7712 replied:
> I think the try catch in persister.readState was added as an additional check in the initial iterations of development

maybe we can move the "additional check" into `persister.readState/persister.writeState`? That can ensure they don't throw exception directly. For example: `validate(request)` could cause exception. 

> we don't seem to need it now. If we're all aligned, then I'll make this change and re-raise a review.

yes, we don't need the `try-catch` if `persister.readState/persister.writeState` always propagate exception by `CompletableFuture`

adixitconfluent replied:
@chia7712 @apoorvmittal10 , I've moved the `try-catch` check  from `SharePartition persister.readState/persister.writeState` to `DefaultStatePersister readState/writeState` methods. Had a discussion with @smjn regarding the scope of try catch block within `readState/writeState` methods, `validate(request)` could only throw such exception. Other exceptions were already being handled by the returned future. Please review when you can. Thanks!

adixitconfluent committed:
Addressed Chia's comment around having try catch in writeState RPC call to handle runtime exceptions

adixitconfluent committed:
Addressed Chia's commentfor removing try catch in writeState/readState RPC call to handle runtime exceptions

adixitconfluent committed:
Addressed Chia's comment for adding try catch in DSP writeState/readState methods to handle runtime exceptions

smjn commented on share/src/main/java/org/apache/kafka/server/share/persister/DefaultStatePersister.java:
```diff
@@ -79,7 +79,12 @@ public CompletableFuture<InitializeShareGroupStateResult> initializeState(Initia
      * @return A completable future of WriteShareGroupStateResult
      */
     public CompletableFuture<WriteShareGroupStateResult> writeState(WriteShareGroupStateParameters request) throws IllegalArgumentException {
```
shall we remove checked exception from the method signature? Also update NoOpPersister signature and tests as well as Persister interface.

smjn commented on share/src/main/java/org/apache/kafka/server/share/persister/DefaultStatePersister.java:
```diff
@@ -170,7 +175,12 @@ WriteShareGroupStateResult writeResponsesToResult(
      * @return A completable future of ReadShareGroupStateResult
      */
     public CompletableFuture<ReadShareGroupStateResult> readState(ReadShareGroupStateParameters request) throws IllegalArgumentException {
```
same here

smjn reviewed (changes requested):
minor comments

adixitconfluent committed:
Addressed Sushant's minor comments

adixitconfluent committed:
Addressed Sushant's minor comment update

smjn reviewed (approved):
LGTM

apoorvmittal10 commented on core/src/test/java/kafka/server/share/SharePartitionTest.java:
```diff
@@ -771,24 +768,20 @@ public void testMaybeInitializeWithReadException() {
         // Complete the future exceptionally for read state.
         Mockito.when(persister.readState(Mockito.any())).thenReturn(FutureUtils.failedFuture(new RuntimeException("Read exception")));
-        SharePartition sharePartition = SharePartitionBuilder.builder().withPersister(persister).build();
+        SharePartition sharePartition1 = SharePartitionBuilder.builder().withPersister(persister).build();
```
nitL why do you need change in varibale names?

adixitconfluent replied:
I find the code cleaner this way since the 2 instances of `SharePartition` behave differently, hence the name change.

apoorvmittal10 reviewed (approved):
Thanks for the changes, LGTM. Just left a minor comment to understand reasoning.

adixitconfluent commented:
@chia7712 , can you please review the changes? Thanks!!

chia7712 commented on share/src/main/java/org/apache/kafka/server/share/persister/Persister.java:
```diff
@@ -40,15 +40,15 @@ public interface Persister {
      * @param request Request parameters
      * @return A {@link CompletableFuture} that completes with the result.
      */
-    CompletableFuture<WriteShareGroupStateResult> writeState(WriteShareGroupStateParameters request) throws IllegalArgumentException;
```
Could you please open a follow-up issue to align exception propagation? Currently, the methods `deleteState`, `initializeState`, and `readSummary` still declare exceptions in their method signatures.

adixitconfluent replied:
thanks @chia7712, created issue https://issues.apache.org/jira/browse/KAFKA-18148, I think we can target it for 4.1 since these APIs are not implemented yet in `DefaultStatePersister`

chia7712 reviewed (approved):
LGTM (again)
